crypto/crypto.go

```diff
@@ -4,75 +4,77 @@ import (
+func ProtectMemory(data []byte) {
+	err := unix.Mlock(data)
+	if err != nil && mlock {
+		// It failed once, probably won't work next time. Supress further warnings.
+		mlock = false
+		fmt.Printf("[!] Warning: Could not mlock() sensitive data; it might get written to SWAP [%s]", err)
 	}
-
-	// Return the CSPR bytes.
-	return b
 }
 
 // Encrypt takes a plaintext and a 32 byte key, encrypts the plaintext with
 // said key using xSalsa20 with a Poly1305 MAC, and returns the ciphertext.
-func Encrypt(plaintext []byte, key []byte) []byte {
+func Encrypt(plaintext []byte, key *[32]byte) ([]byte, error) {
 	// Generate a random nonce.
-	nonceSlice := generateRandomBytes(24)
+	nonceSlice, err := generateRandomBytes(24)
+	if err != nil {
+		return nil, err
+	}
 
 	// Store it in an array.
 	var nonce [24]byte
 	copy(nonce[:], nonceSlice)
 
-	// Store the symmetric key in an array.
-	var secretKey [32]byte
-	copy(secretKey[:], key)
-
-	// Encrypt the plaintext.
-	ciphertext := secretbox.Seal(nonce[:], plaintext, &nonce, &secretKey)
-
-	// Return the base64 encoded ciphertext.
-	return ciphertext
+	// Encrypt and return the plaintext.
+	return secretbox.Seal(nonce[:], plaintext, &nonce, key), nil
 }
 
 // Decrypt takes a ciphertext and a 32 byte key, decrypts the ciphertext with
-// said key, and then returns the plaintext. If the key is incorrect, decryption
-// fails and the program terminates with exit code 1.
-func Decrypt(ciphertext []byte, key []byte) []byte {
+// said key, and then returns the plaintext.
+func Decrypt(ciphertext []byte, key *[32]byte) ([]byte, error) {
 	// Grab the nonce from the ciphertext and store it in an array.
 	var nonce [24]byte
 	copy(nonce[:], ciphertext[:24])
 
-	// Store the symmetric key in an array.
-	var secretKey [32]byte
-	copy(secretKey[:], key)
-
 	// Decrypt the ciphertext and store the result.
-	plaintext, okay := secretbox.Open([]byte{}, ciphertext[24:], &nonce, &secretKey)
+	plaintext, okay := secretbox.Open([]byte{}, ciphertext[24:], &nonce, key)
 	if !okay {
-		fmt.Println("[!] Decryption failed")
-		os.Exit(1)
+		// This shouldn't happen.
+		return nil, errors.New("[!] Decryption of data failed")
 	}
 
+	// Protect the plaintext.
+	ProtectMemory(plaintext)
+
 	// Return the resulting plaintext.
-	return plaintext
+	return plaintext, nil
 }
 
 // DeriveKey derives a 32 byte encryption key from a password and identifier.
-func DeriveKey(password, identifier []byte, cost map[string]int) []byte {
-	derivedKey, _ := scrypt.Key(password, identifier, 1<<uint(cost["N"]), cost["r"], cost["p"], 32)
-	return derivedKey
+func DeriveKey(password, identifier []byte, cost map[string]int) *[32]byte {
+	derivedKeySlice, _ := scrypt.Key(password, identifier, 1<<uint(cost["N"]), cost["r"], cost["p"], 32)
+	ProtectMemory(derivedKeySlice)
+
+	// Convert to fixed-size array.
+	var derivedKey [32]byte
+	copy(derivedKey[:], derivedKeySlice)
+	ProtectMemory(derivedKey[:])
+
+	// Return a pointer.
+	return &derivedKey
```

Comment:
I'd suggest declaring `derivedKey` as a pointer, then de-referencing on the `copy()`, and just returning the pointer